### DISCUSSION TITLE
Requesting your feedback - Jotai DevTools Alpha preview

### DISCUSSION THREAD

**Discussion by arjunvegda** [pmndrs/jotai#1754] [General]
We've made some great progress on the Alpha version of Jotai UI DevTools 🎉 

And we'd love to gather user opinions and suggestions in order to improve its functionality and user experience 🙌

#### ✨ Preview
![image](https://user-images.githubusercontent.com/14841132/217137308-c2761642-f2b2-4dc5-a8ce-ee685f556268.png)

#### 🙌 Process

1. Find up-to-date docs on the [`jotai-devtools` repo's `ui-devtools` branch](https://github.com/jotaijs/jotai-devtools/tree/ui-devtools#readme)
2. Access the Alpha version of Jotai UI DevTools via this [CodeSandbox link](https://k5p12d.csb.app/)
3. Play around with the tool and familiarize yourself with its features
4. Provide your feedback in this discussion thread below in the form of suggestions, bug reports, or any other observations you may have

#### 🗒️ Note
You are welcome to fork the [CodeSandbox](https://k5p12d.csb.app/) to make changes and see how they affect the tool. 

Or you can just follow the instructions below to test it on your own project 🎉

#### 📦 Add it to your own project

More in-depth instructions can be found [here](https://github.com/jotaijs/jotai-devtools/tree/ui-devtools#readme) 

```sh
# yarn
yarn add jotai-devtools@next

# npm
npm install jotai-devtools@next --save
```

Follow these [instructions](https://github.com/jotaijs/jotai-devtools/tree/ui-devtools#babel-plugin-setup---optional-but-highly-recommended) to set up the Babel plugins (Optional but highly recommended)

**Comment by barelyhuman** [comment]
That looks amazing! 

Few things, I'd like to address though 

1. Is it possible to show a visual difference between a normal atom and a dependent atom. As in, right now all atoms are shown in the same way on the left list, could there be a visual cue to show that something is a dependent atom like a tree or just a badge to start with?
2. Any plans of maintaining historical value? I know it's not easy to do right now, I just wanted to know if we were even planning to do that

**Reply by arjunvegda** [reply]
Thanks for sharing your thoughts! 

> Is it possible to show a visual difference between a normal atom and a dependent atom. As in, right now all atoms are shown in the same way on the left list, could there be a visual cue to show that something is a dependent atom like a tree or just a badge to start with?

That's a good idea! I'm planning to spin up a second tab that helps users visualize atom graphs. Here is a rough sketch [from a while ago](https://github.com/pmndrs/jotai/discussions/1637#discussioncomment-4568876) 👇
 
<img width="700" alt="image" src="https://user-images.githubusercontent.com/14841132/217370069-8c93d3dd-8e23-4310-846d-80b2fc39dd73.png">

What do you think? Would this suffice?

> Any plans of maintaining historical value? I know it's not easy to do right now, I just wanted to know if we were even planning to do that

Yes! I'm planning to add the "Time Travel" capabilities (similar to Redux DevTools). Here is a rough sketch [from a while ago](https://github.com/pmndrs/jotai/discussions/1637#discussioncomment-4568876) 👇

<img width="700" alt="image" src="https://user-images.githubusercontent.com/14841132/217370878-9394e1c7-a76b-4a34-a6bf-5218d4250b4f.png">

Any thoughts on it?

**Reply by barelyhuman** [reply]
The graph seems amazing but i was more on the tree side of things. You’ve mentioned a few examples in replies below. 

Yeah the time line looks great. Will be waiting for that 👀

**Reply by arjunvegda** [reply]
Awesome! yeah, I'm open to exploring the tree structure approach. I'll let you know once I start looking into it.

**Comment by t6adev** [comment]
Hi! Thank you for your great work!
Does the current version show all atoms, right?
It's just my hopeful imagination. It's good for me that some atoms on any one component will be displayed.

Now I'm trying the jotai-location combination.
[codesandbox](https://codesandbox.io/s/jotai-devtools-demo-with-location-znvlh4?file=/src/App.tsx)
This is so nice because only mounted atoms are shown. My suggestion is to show atoms are each component.

Anyway, in this progress, I realize base atoms are displayed as `<unlabeled-atom>`.
Those atoms' dependants are known so I hope they could be displayed with grouping styling, like with indent or accordion.
![image](https://user-images.githubusercontent.com/6883692/217405409-5fa41858-d260-4c3f-836a-519c9c5859ed.png)

Oh, yes, atoms could be grouped by mounted component? If it does, it's also a nice way.

**Reply by arjunvegda** [reply]
Thanks a lot for giving it a try and sharing your feedback!

> Does the current version show all atoms, right?
It's just my hopeful imagination. It's good for me that some atoms on any one component will be displayed.

Yes, I've tested almost all the primitives and it should work as expected. Please let me know if you find any anomalies.

> Anyway, in this progress, I realize base atoms are displayed as <unlabeled-atom>.

Great feedback! Daishi and I discussed the idea of introducing a new `internal` property on atoms that we could use to mark atoms as "internal" explicitly a while ago. 

There are instances where debugging those internal atoms might be needed so we might make it configurable at the `<DevTools/>` level.

I'll look into it in the coming weeks and keep you posted. 🙌

> Those atoms' dependants are known so I hope they could be displayed with grouping styling, like with indent or accordion.

Interesting idea! So a tree-like structure for deeply nested dependents?
Eg. 
```
├─ atom A
│  ├─ atom A-1
│  ├─ atom A-2
├─ atom B
```

> Oh, yes, atoms could be grouped by mounted component? If it does, it's also a nice way.

That'd be cool! Although, I wonder if this could lead to more overhead both cognitive (on the user's end) and on maintainers since it'd require quite a bit of parsing. I'll keep this one on the back burner for now.

**Reply by t6adev** [reply]
> So a tree-like structure for deeply nested dependents?

Yes!!
I'm not sure which hierarchy pattern is a better way, it means that dependant atoms should be displayed as parents or dependency atoms should be.

When we'd like to check the siblings of a parent atom, it might work well.
```
├─ base atom A
│  ├─ derived atom A-1
│  ├─ derived atom A-2
```

or
When we'd like to check just leaves, it might work well.
```
├─ derived atom A-1
│  ├─ base atom A
├─ derived atom A-2
│  ├─ base atom A
```

It might be good to provide a toggle feature.

**Reply by arjunvegda** [reply]
Great! I'll put this in my backlog and let you know once I get a chance to look into it deeply. Thanks again for sharing your ideas 🙌

**Reply by arjunvegda** [reply]
Hey @t6adev

A small update for you.

> Anyway, in this progress, I realize base atoms are displayed as .
 Great feedback! Daishi and I discussed the idea of introducing a new internal property on atoms that we could use to mark atoms as "internal" explicitly a while ago.

I've addressed this feedback 👆. Users can now configure the display of private atoms via `options.shouldShowPrivateAtoms` prop. Defaults to `false`. 

Here is the CodeSandbox with the fixed version 🙌 Let me know what you think
https://codesandbox.io/s/jotai-devtools-demo-with-location-forked-xrtdwh?file=/src/App.tsx

**Comment by MiroslavPetrik** [comment]
Let's use some colors for the 'Raw value' section:
![image](https://user-images.githubusercontent.com/38855190/217654020-c4ac36a2-f809-4e09-a76e-fa894b1160ee.png)

Also when value is string, it should have quotes around:
`"hello"`

Second idea: the r aw value could have copy to clipboard on the right side, like github has with code block:

```js
"ff780375-c426-4ca3-8b6b-cb90270702c3"
```

**Reply by arjunvegda** [reply]
Great feedback. Thanks for sharing!

> Let's use some colors for the 'Raw value' section:

Good idea! Colors are really hard to get right, I was playing around with something like this last week (haven't pushed it yet). I plan on coming back to this as part of my UI/UX overhaul once I flush some core features out. I'll keep ya posted 🙌 

<img width="428" alt="image" src="https://user-images.githubusercontent.com/14841132/217711218-8745c899-6722-496d-936d-23daa31a7bdb.png">

> Also when value is string, it should have quotes around:

Thanks for the pointer! Fixed it in the canary release. Here is the [link](https://k5p12d.csb.app/) to the updated CSB

>Second idea: the r aw value could have copy to clipboard on the right side, like github has with code block:

Seems easy enough! Fixed it in the canary release. Here is the [link](https://k5p12d.csb.app/) to the updated CSB

**Reply by MiroslavPetrik** [reply]
KING 🦍
I see that `null` is already in red color.
I like orange strings, the number in greenish blue, and for booleans IDK.

also I'm on windows chrome and the atom names don't have font (compared to your original screenshot above):
![image](https://user-images.githubusercontent.com/38855190/217784583-bc5b48da-d5c3-4158-ac93-6f3dfdcc9e31.png)

**Reply by arjunvegda** [reply]
> also I'm on windows chrome and the atom names don't have font (compared to your original screenshot above):

This appears to be CSB issue, it doesn't load fonts nicely. Is this happening locally too? Could you help us test? 
![image](https://user-images.githubusercontent.com/14841132/217848554-caaaa00d-c6b6-445a-be7f-39e81b0b6f93.png)

**Reply by MiroslavPetrik** [reply]
I'm running it in storybook and it looks the same:
![image](https://user-images.githubusercontent.com/38855190/219949474-8fab76bd-560c-4b33-8151-a55f0e7b2e3e.png)

Perhaps some styles need to be imported?

**Reply by arjunvegda** [reply]
~Thank you for checking! We made some improvements to how we managed CSS. May I ask you to give `jotai-devtools@0.3.0-next.3` a shot?~

~Here is a link to the updated CSB - https://k5p12d.csb.app/~

Never mind, the issue still persists. I managed to reproduce it on my end. I'll continue my investigation.

**Reply by arjunvegda** [reply]
@MiroslavPetrik - Circling back to let you know that the font issue should be fixed now, it was caused by CRA😅. May I ask you to give `jotai-devtools@next` a shot?

**Comment by MiroslavPetrik** [comment]
I'm curious whether some debugLabel patterns could be established, by which these dev tools would be able to highlight, group, or filter atoms.

For example I have certain atoms which represent input value, and others which represent validation state.

since all of my inputs can have `name`, I could establish pattern for the debug labels like so

- `[inputValueAtom]/${name}`
- `[validationStateAtom]/${name}`

The 'atom kinds' in the brackets could be then used by the dev tools for groupings. Yes, currently I can filter all the values with `inputValueAtom`, but eventually, more advanced uses, like _tree list_ rendering could be used.

I imagine that namespacing like this would be useful to distinguish state between 
a) different libraries
b) different features

**Reply by arjunvegda** [reply]
This is a great point! I like the idea of `namespaces` for debug labels.  

Perhaps we could introduce something like `debugNamespaces` 🤔 
```js
const inputAtomA = atom();
inputAtomA.debugNamespaces = ['inputValueAtom'];

const validationStateAtomA = atom();
validationStateAtomA.debugNamespaces = ['validationStateAtom'];
```

I'll put this in my backlog and revisit this after wrapping up some P0 stability features.

**Comment by AjaxSolutions** [comment]
I have two stores in my app. I'd like to be able to view and search atoms in each store.  How can I do this?  I don't see that the atom viewer allows me to select a store.

**Reply by arjunvegda** [reply]
You're right! We don't support multiple stores at this point. I'll put it in my backlog as a low priority for now. I haven't seen a lot of people using multiple stores so if there is enough demand, I can bump it up in priority 🙌

**Reply by dai-shi** [reply]
I'd see it as rather a high priority because it's a fundamental concept in Jotai. It doesn't have to be a nice UI, but somehow it should support multiple stores even in a naive way.

**Reply by arjunvegda** [reply]
That makes sense! I could bump it up in priority after the initial first stable release. I'll keep you posted on the progress @AjaxSolutions 🙌
